packages/chatbot-server-mongodb-public/src/test/testHelpers.ts

```diff
@@ -45,4 +45,32 @@ export async function makeTestApp(defaultConfigOverrides?: Partial<AppConfig>) {
   };
 }
 
-export { systemPrompt, generateUserPrompt } from "../config";
+export function cosineSimilarity(a: number[], b: number[]) {
+  // https://towardsdatascience.com/how-to-build-a-textual-similarity-analysis-web-app-aa3139d4fb71
+
+  const magnitudeA = Math.sqrt(dotProduct(a, a));
+  const magnitudeB = Math.sqrt(dotProduct(b, b));
+  if (magnitudeA && magnitudeB) {
+    // https://towardsdatascience.com/how-to-measure-distances-in-machine-learning-13a396aa34ce
+    return dotProduct(a, b) / (magnitudeA * magnitudeB);
+  } else {
+    return 0;
+  }
+}
+
+function dotProduct(a: number[], b: number[]) {
+  let sum = 0;
```

Comment:
Nit: in a reversal of one of your reviews of mine, I think you could stand to get comfortable with `Array.reduce()` and rewrite this as:
```js
return a.reduce((acc, cur, i) => acc + cur * b[i], 0);
```

Reply:
def (this implementation was a copy-pasta)